### DISCUSSION TITLE
Create a ContextMenu from xaml, per DataGrid row

### DISCUSSION THREAD

**Discussion by Dragorn421** [AvaloniaUI/Avalonia#16042] [Q&A] [answered]
Hi,

I'm using a DataGrid and I want a different ContextMenu on each row (e.g. depending on the row's DataContext).

My current solution is using `LoadingRow="OnDataGridLoadingRow"` on the `DataGrid` element in the xaml, and in the `OnDataGridLoadingRow` function setting the `(e as DataGridRowEventArgs).Row.ContextMenu` to a new code-constructed `ContextMenu` instance:
```cs
    public void OnDataGridLoadingRow(object? sender, DataGridRowEventArgs e)
    {
        var cm = new ContextMenu();
        var ri = e.Row.DataContext as MainWindowViewModel.RowItem;
        cm.Items.Add(new MenuItem() { Header = "thing " + ri.ValueCol1 });
        cm.Items.Add(new MenuItem() { Header = "hey" });
        e.Row.ContextMenu = cm;
    }
```

This works, but I would like to define the `ContextMenu` in xaml as much as possible instead of having to create it and its items in code.

Here is minimal full code
https://github.com/Dragorn421/DragoStuff/tree/avalonia_contextmenu_xaml
showing
- Creating a `ContextMenu` and populating it with items in code, leading to the expected result (`MainWindow.OnDataGridLoadingRow`, upper datagrid in the window)
- Trying to create a xaml-based context menu `ContextMenuTemplate` (doesn't work, it does build and run without errors or any warning in the console output, but no context menu shows) (`MainWindow.OnDataGridWTemplateAttemptLoadingRow`, lower datagrid in the window)

Is there a better approach than using `LoadingRow=`?
Can a ContextMenu be created from xaml?

**Comment by timunie** [comment, marked as answer]
You can add a Style for `DataGridRow` which hold a ContextFlyout iirc.

**Reply by Dragorn421** [reply]
Thanks, using a Style selector to set DataGridRow's ContextMenu works :)
Only hurdle is `x:DataType` needs to be set on the `ContextMenu` to the rows' DataContext type

```xaml
<DataGrid
    ItemsSource="{Binding MyDataGridItems}"
>
    <DataGrid.Styles>
        <Style Selector="DataGridRow">
            <Setter Property="ContextMenu">
                <ContextMenu x:DataType="vm:MainWindowViewModelRowItem">
                    <MenuItem Header="{Binding ValueCol1}" />
                    <ContextMenu.Resources>
                        <vm:MyConverter x:Key="myConverter"/>
                    </ContextMenu.Resources>
                    <MenuItem Header="{Binding ValueCol1,
                                        Converter={StaticResource myConverter},
                                        ConverterParameter='a prefix '}" />
                    <MenuItem Header="runthing" Command="{Binding RunThing}" />
                    <MenuItem Header="heyy" />
                </ContextMenu>
            </Setter>
        </Style>
    </DataGrid.Styles>
    <DataGrid.Columns>
        <DataGridTextColumn Header="Col1" Binding="{Binding ValueCol1}" />
    </DataGrid.Columns>
</DataGrid>
```

pushed updated example to https://github.com/Dragorn421/DragoStuff/tree/avalonia_contextmenu_xaml
